File: 1250-longest-common-subsequence/1250-longest-common-subsequence.py
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        l2 = len(text2)
        prev_row = [0] * (l2+1)
        curr_row = [0] * (l2+1)

        for s in text1:
            for i in range(1, l2 + 1):
                if s == text2[i-1]:
                    curr_row[i] = prev_row[i-1] + 1
                else:
                    curr_row[i] = max(prev_row[i], curr_row[i-1])
            prev_row, curr_row = curr_row, prev_row

        return prev_row[-1]

        

File: 1250-longest-common-subsequence/1250-longest-common-subsequence.js
/**
 * @param {string} text1
 * @param {string} text2
 * @return {number}
 */
var longestCommonSubsequence = function(text1, text2) {
    const dp = new Array(text1.length + 1).fill().map(() => new Array(text2.length + 1).fill(0))
    // need extra space for base case
    for(let i = text1.length - 1; i >= 0; i--){
        for(let j = text2.length - 1; j >= 0; j--){
            if(text1[i] === text2[j]){
                dp[i][j] = 1 + dp[i + 1][j + 1]
            } else {
                dp[i][j] = Math.max(dp[i][j + 1], dp[i + 1][j])
            }
        }
    }
    return dp[0][0]
};
// if char matching, get diagonal value + 1
// if not matching, max of right and bottom